From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mio <mio@mio19.uk>
Date: Thu, 22 Oct 2026 01:13:05 +1100
Subject: [PATCH] examples/asteroid: align launcher capabilities, polkit, and
 nightly pin

Grant the hwcomposer capability, install the upstream polkit rule, and
pin the launcher to the current nightly that uses it.

Signed-off-by: Mio <mio@mio19.uk>
---
 examples/asteroid/asteroid-desktop.nix        | 26 ++++++++++++++++++-
 .../asteroidos/asteroid-launcher/default.nix  |  4 +--
 2 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/examples/asteroid/asteroid-desktop.nix b/examples/asteroid/asteroid-desktop.nix
index 7ca898e0..c30b1c67 100644
--- a/examples/asteroid/asteroid-desktop.nix
+++ b/examples/asteroid/asteroid-desktop.nix
@@ -150,12 +150,36 @@ in
       }
       // config.environment.variables;
       serviceConfig = {
-        ExecStart = "${pkgs.asteroid.asteroid-launcher}/bin/asteroid-launcher $LIPSTICK_OPTIONS -plugin evdevtouch --systemd";
+        ExecStart = "${config.security.wrapperDir}/asteroid-launcher $LIPSTICK_OPTIONS -plugin evdevtouch --systemd";
         Restart = "always";
         RestartSec = "2s";
       };
     };
 
+    # Upstream AsteroidOS setcap's the launcher with cap_wake_alarm and
+    # cap_block_suspend for always-on-display ticks; expose the same file
+    # capabilities through the NixOS wrapper.
+    # Reference: https://github.com/AsteroidOS/meta-asteroid/blob/master/recipes-asteroid/asteroid-launcher/asteroid-launcher_git.bb
+    security.wrappers.asteroid-launcher = {
+      source = "${pkgs.asteroid.asteroid-launcher}/bin/asteroid-launcher";
+      capabilities = "cap_wake_alarm,cap_block_suspend+ep";
+      owner = "root";
+      group = "root";
+    };
+
+    # Upstream AsteroidOS allows the ceres user to reboot and power off.
+    # Reference: https://github.com/AsteroidOS/meta-asteroid/blob/master/recipes-asteroid/asteroid-apps/files/30-org.freedesktop.login1.rules
+    security.polkit.extraConfig = ''
+      polkit.addRule(function(action, subject) {
+        if ((action.id == "org.freedesktop.login1.set-reboot-parameter" ||
+             action.id == "org.freedesktop.login1.power-off-multiple-sessions" ||
+             action.id == "org.freedesktop.login1.reboot-multiple-sessions") &&
+            subject.user == "ceres") {
+          return polkit.Result.YES;
+        }
+      });
+    '';
+
     # AsteroidOS Bluetooth synchronization daemon service
     systemd.services.asteroid-btsyncd = {
       description = "AsteroidOS Bluetooth Sync Daemon";
diff --git a/overlay/asteroidos/asteroid-launcher/default.nix b/overlay/asteroidos/asteroid-launcher/default.nix
index 921afd1b..1a58d148 100644
--- a/overlay/asteroidos/asteroid-launcher/default.nix
+++ b/overlay/asteroidos/asteroid-launcher/default.nix
@@ -30,8 +30,8 @@ stdenv.mkDerivation (finalAttrs: {
   src = fetchFromGitHub {
     owner = "AsteroidOS";
     repo = "asteroid-launcher";
-    rev = "7eae1ebc86b04377cdb3164a8027a1667de319a2";
-    sha256 = "0inra188hizfap94dc95a13q1c82bqc5dsa6qdgkdsvim65fg3vq";
+    rev = "4bd3d281fe3a2ea9631c640e1f00871fa19bbd4c";
+    sha256 = "194hn7pkq100339gnavvjmyib15m557bykkyqr77jrgz0wl07bpz";
   };
 
   strictDeps = true;
